packages/flutter/lib/src/widgets/localized_resources.dart

```diff
@@ -0,0 +1,354 @@
+///    which enable configuring the app's LocalizedResourcesLoader.
+class DefaultLocalizedResourcesDelegate extends LocalizedResourcesDelegate {
+  DefaultLocalizedResourcesDelegate(this.allLoaders) {
+    assert(allLoaders != null);
+  }
+
+  final Map<Type, LocalizedResourceLoader> allLoaders;
+
+  final Map<Locale, Map<Type, dynamic>> _localeToResources = <Locale, Map<Type, dynamic>>{};
+  final Set<Locale> _loading = new Set<Locale>();
+
+  @override
+  Future<Null> load(Locale locale) {
+    assert(locale != null);
+    assert(!_loading.contains(locale));
+    final Map<Type, dynamic> resourceMap = <Type, dynamic>{};
+
+    _loading.add(locale);
+
+    // This is more complicated then just making a list of allLoaders[type](locale)
+    // for each type in allTypes and then Future.wait-ing for all of the futures
+    // because some of the loaders may return SynchronousFutures. We don't want
+    // to Future.wait for the synchronous futures.
+    Map<Type, Future<dynamic>> resourceMapFutures;
+    for(Type type in allLoaders.keys) {
+      dynamic completedValue;
+      final Future<dynamic> futureValue = allLoaders[type](locale).then<dynamic>((dynamic value) {
+        return completedValue = value;
+      });
+      if (completedValue != null) {
+        resourceMap[type] = completedValue;
+      } else {
+        resourceMapFutures ??= <Type, Future<dynamic>>{};
+        resourceMapFutures[type] = futureValue;
+      }
+    }
+
+    // Common case: only the toolkit resources were loaded and they were all
+    // synchronous futures.
+    if (resourceMapFutures == null) {
+      _loading.remove(locale);
+      _localeToResources[locale] = resourceMap;
+      return new SynchronousFuture<Null>(null);
+    }
+
+    // Wait on the remaining futures and update resourceMap.
+    return Future.wait<Null>(resourceMapFutures.values,
+      eagerError: true,
+      cleanUp: (dynamic value) {
+        _loading.remove(locale);
+      }
+    ).then((List<dynamic> allValues) {
+      final List<Type> allTypes = resourceMapFutures.keys.toList();
+      _loading.remove(locale);
+      for (int i = 0; i < allTypes.length; i += 1) {
+        assert(allValues[i] != null);
+        resourceMap[allTypes[i]] = allValues[i];
+      }
+      _localeToResources[locale] = resourceMap;
+      return null;
+    });
+  }
+
+  @override
+  LocalizedResourcesStatus statusFor(Locale locale) {
+    if (_localeToResources[locale] != null)
+      return LocalizedResourcesStatus.loaded;
+    if (_loading.contains(locale))
+      return LocalizedResourcesStatus.loading;
+    return LocalizedResourcesStatus.none;
+  }
+
+  @override
+  T resourcesFor<T>(Locale locale, Type type) {
+    assert(locale != null);
+    assert(type != null);
+    final Map<Type, dynamic> resources = _localeToResources[locale];
+    if (resources == null)
+      return null;
+    // TBD: assert(resources[type].runtimeType, type)? Too restrictive?
```

Comment:
yeah, what you really want is `resources[type] is type` but for that you need https://github.com/dart-lang/sdk/issues/27680.